### PR TITLE
added achievements for the guards

### LINKED ISSUE #226: Guard Achievements
Add achievements for building guard towers to lvl 1 and max.

Maybe add some for killing x mobs or for seasoned veterans.

### PR DESCRIPTION
**Pull request by strwbrryfrsh**
Closes #226 

# Changes proposed in this pull request:
- added achievement for killed mobs
- added achievement for guard tower

Review please @Minecolonies/maintainers

### CODE REVIEW

File: src/main/java/com/minecolonies/entity/ai/citizen/guard/AbstractEntityAIGuard.java [hunk 1]
```diff
@@ -1,5 +1,8 @@
 package com.minecolonies.entity.ai.citizen.guard;
 
+import com.blockout.Log;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/guard/AbstractEntityAIGuard.java]: MINOR Remove this unused import 'com.blockout.Log'. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/minecolonies/entity/ai/citizen/guard/AbstractEntityAIGuard.java [hunk 2]
```diff
@@ -1,5 +1,8 @@
...
+import com.minecolonies.achievements.ModAchievements;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/guard/AbstractEntityAIGuard.java]: MINOR Remove this unused import 'com.minecolonies.achievements.ModAchievements'. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/minecolonies/entity/ai/citizen/guard/AbstractEntityAIGuard.java [hunk 3]
```diff
@@ -331,6 +334,17 @@ private BlockPos getRandomBuilding()
     }
 
     /**
+     * Called when a guard killed an entity.
+     *
+     * @param killedEntity the entity being killed.
+     */
+    protected void onKilledEntity(EntityLivingBase killedEntity)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/guard/AbstractEntityAIGuard.java]: MAJOR Remove this unused method parameter "killedEntity". [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1172)

File: src/main/java/com/minecolonies/entity/ai/citizen/guard/AbstractEntityAIGuard.java [hunk 4]
```diff
@@ -331,6 +334,17 @@ private BlockPos getRandomBuilding()
...
+    {
+        Colony colony = this.getOwnBuilding().getColony();
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/guard/AbstractEntityAIGuard.java]: MINOR Local variable 'colony' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/entity/ai/citizen/guard/EntityAIMeleeGuard.java [hunk 5]
```diff
@@ -1,5 +1,6 @@
 package com.minecolonies.entity.ai.citizen.guard;
 
+import com.blockout.Log;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/guard/EntityAIMeleeGuard.java]: MINOR Remove this unused import 'com.blockout.Log'. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/minecolonies/entity/ai/citizen/guard/GuardArrow.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+/**
+ * Arrow class for arrows shot by guards.
+ */
+public class GuardArrow extends EntityTippedArrow
+{
+    private static final String TAG_COLONY = "colony";
+    private Colony colony;
+
+    public GuardArrow(final World worldin)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/guard/GuardArrow.java]: MINOR Document this public constructor. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUndocumentedApi)

File: src/main/java/com/minecolonies/entity/ai/citizen/guard/GuardArrow.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    {
+        super(worldin);
+    }
+
+    public GuardArrow(final World worldIn, final EntityCitizen shooter)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/guard/GuardArrow.java]: MINOR Document this public constructor. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUndocumentedApi)

File: src/main/java/com/minecolonies/entity/ai/citizen/guard/GuardArrow.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    {
+        super.writeEntityToNBT(compound);
+        compound.setInteger(TAG_COLONY, colony.getID());
+    }
+
+    @Override
+    public void readEntityFromNBT(final NBTTagCompound compound)
+    {
+        super.readEntityFromNBT(compound);
+        int colonyID = compound.getInteger(TAG_COLONY);
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/guard/GuardArrow.java]: MINOR Local variable 'colonyID' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/achievements/ModAchievements.java [hunk 9]
```diff
@@ -136,6 +137,45 @@
...
+     * Kill 25 mobs
+     */
+    public static final Achievement achievementKill25Mobs = new MineColoniesAchievement("guard.mobkill.25", 4, 8, Items.ROTTEN_FLESH, achievementKillOneMob).registerStat();
+
+    /**
+     * Kill 100 mobs
+     */
+    public static final Achievement achievementKill100Mobs = new MineColoniesAchievement("guard.mobkill.100", 6, 8, Items.GUNPOWDER, achievementKill25Mobs).registerStat();
+
+    /**
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/achievements/ModAchievements.java]: MINOR First sentence should end with a period. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=checkstyle%3AJavadoc_Style)

File: src/main/java/com/minecolonies/achievements/ModAchievements.java [hunk 10]
```diff
@@ -136,6 +137,45 @@
...
+     */
+    public static final Achievement achievementBuildingGuard   = new MineColoniesAchievement("upgrade.guard.first", 0, 7, ModBlocks.blockHutGuardTower,
+                                                                                               achievementTownhall).registerStat();
+    /**
+     * Max out a guard.
+     */
+    public static final Achievement achievementUpgradeGuardMax = new MineColoniesAchievement("upgrade.guard.max", 2, 7, ModBlocks.blockHutGuardTower,
+                                                                                               achievementBuildingGuard).registerStat();
+
+    /**
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/achievements/ModAchievements.java]: MINOR First sentence should end with a period. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=checkstyle%3AJavadoc_Style)

File: src/main/java/com/minecolonies/achievements/ModAchievements.java [hunk 11]
```diff
@@ -136,6 +137,45 @@
...
+     * Kill one mob
+     */
+    public static final Achievement achievementKillOneMob = new MineColoniesAchievement("guard.mobkill.one", 2, 8, Items.BONE, achievementBuildingGuard).registerStat();
+
+    /**
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/achievements/ModAchievements.java]: MINOR First sentence should end with a period. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=checkstyle%3AJavadoc_Style)

File: src/main/java/com/minecolonies/colony/Colony.java [hunk 12]
```diff
@@ -351,6 +356,47 @@ private void markDirty()
...
+        }
+        if (mobKills >= 100)
+        {
+            this.triggerAchievement(ModAchievements.achievementKill100Mobs);
+        }
+        if (mobKills >= 500)
+        {
+            this.triggerAchievement(ModAchievements.achievementKill500Mobs);
+        }
+        if (mobKills >= 1000)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/colony/Colony.java]: MINOR Assign this magic number 1000 to a well-named constant, and use the constant instead. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS109)

File: src/main/java/com/minecolonies/colony/Colony.java [hunk 13]
```diff
@@ -351,6 +356,47 @@ private void markDirty()
...
+     */
+    public void incrementMobsKilled()
+    {
+        killedMobs++;
+        final int mobKills = this.getKilledMobs();
+        if (mobKills >= 1)
+        {
+            this.triggerAchievement(ModAchievements.achievementKillOneMob);
+        }
+        if (mobKills >= 25)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/colony/Colony.java]: MINOR Assign this magic number 25 to a well-named constant, and use the constant instead. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS109)

File: src/main/java/com/minecolonies/colony/Colony.java [hunk 14]
```diff
@@ -351,6 +356,47 @@ private void markDirty()
...
+        {
+            this.triggerAchievement(ModAchievements.achievementKill25Mobs);
+        }
+        if (mobKills >= 100)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/colony/Colony.java]: MINOR Assign this magic number 100 to a well-named constant, and use the constant instead. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS109)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 18 issues
* MAJOR 2 major
* MINOR 16 minor

Watch the comments in this conversation to review them.

**Comment by Raycoms** [comment]
:shipit:

**Pull request merged**